src/track/beatutils.cpp

```diff
@@ -254,52 +247,69 @@ QVector<double> BeatUtils::calculateIronedBeatMap(
     double previousPhaseErrorMean = 0.0;
     auto phaseErrorAvarageCalculator = MovingAvarage(1);
     auto bpmAvarageCalculator = MovingAvarage(1);
-    double meanBpm;
+    double meanBpm = 0.0;
+    double meanBeatLength = 0.0;
+    bool foundFistOutlier = false;
     QVector<double> fixedBeats;
-    while (rightIndex < rawbeats.size()) {
+    while (beatOffset < rawbeats.last() + meanBeatLength) {
         double bpm = 60.0 * sampleRate / (rawbeats[rightIndex] - rawbeats[rightIndex -1]);
         meanBpm = bpmAvarageCalculator(bpm);
         bpmAvarageCalculator.increasePeriod();
         // we do the rounding because we do not want beats at fractional sample position
-        double meanBeatLength = floor(((60.0 * sampleRate) / meanBpm) + 0.5);
+        meanBeatLength = floor(((60.0 * sampleRate) / meanBpm) + 0.5);
         phaseError = secondsPerSample * (beatOffset - rawbeats[rightIndex]);
         phaseErrorMean = phaseErrorAvarageCalculator(phaseError);
         phaseErrorAvarageCalculator.increasePeriod();
         isErrorGrowing = checkIfErrorGrowing(previousPhaseErrorMean, phaseErrorMean);
         if (isErrorGrowing && !wasErrorGrowing) {
             errorStartedGrowing = rightIndex;
         }
-        double correctedOffset = rawbeats[lastCorrectedBeat];
-        // we reached our phase error, add the beats until error started growing
+        double correctedOffset;
+        if (lastCorrectedBeat == 0) {
+            correctedOffset = rawbeats[lastCorrectedBeat];
+        } else {
+            correctedOffset = fixedBeats.last();
+        }
         if (fabs(phaseError) > kMaxSecsPhaseError) {
-            qDebug() << "from beat" << lastCorrectedBeat << "to beat" << errorStartedGrowing << "tempo is" << meanBpm;
-            for (int i = lastCorrectedBeat; i < errorStartedGrowing; i += 1) {
-                correctedOffset += meanBeatLength;
-                fixedBeats << correctedOffset;
-            }
-            lastCorrectedBeat = errorStartedGrowing;
-            if (lastCorrectedBeat == rawbeats.size() -1) {
-                break;
+            // ignore a single outlier
+            if (!foundFistOutlier) {
+                foundFistOutlier = true;
+                isErrorGrowing = false;
+            // add the beats until error started growing
+            } else {
+                qDebug() << "from beat" << lastCorrectedBeat << "to beat" << errorStartedGrowing << "tempo is" << meanBpm;
+                for (int i = lastCorrectedBeat; i < errorStartedGrowing; i += 1) {
+                    correctedOffset += meanBeatLength;
```

Comment:
Why are you only go to the "errorStartedGrowing" beat? 
But you use the average until the secondOutlier.
This way the average is not optimal. because it is calculated from the outlier and you do not reach zero error for the current beat.

Reply:
Please change the algorithm to what I have done in the spread sheet.
I have noticed to achieve better results going backwards from a longest region.  

* loop backwards through the raw beats. and calculate the average beat length from the first beat.  
* add an inner loop and check for outliers using the momentary average as  beat length.
* once you have found an average with only single outlyers, store the beats using the current avarage.
* reset and do the loop again, starting with the region from the found beat to the end.

Reply:
Ok, thanks for the clear explanation. I think I finally got it.

Reply:
> Why are you only go to the "errorStartedGrowing" beat?
> But you use the average until the secondOutlier.
> This way the average is not optimal. because it is calculated from the outlier and you do not reach zero error for the current beat.

The idea is to use it to avoid to reduce a growing offset instead of arbitrarily split the region into four like you originally suggested a few days ago.
I know you just request me to implemented the latest algorithm on the spreadsheet and I agreed, but I made a simple two line change to this algorithm to use the average only until errorStartedGrowing like you pointed and I think it now works really well. IMHO maybe the curve looks even better (more changes, but smaller jumps) and I did not notice any phase problems on the beatgrid.

Yellow curve is the next commit.

![newIroning](https://user-images.githubusercontent.com/61819301/91373152-03c8a280-e7ec-11ea-8508-662eb86d9fab.png)